rust/lance/src/dataset/optimize/remapping.rs

```diff
@@ -176,6 +181,152 @@ pub fn transpose_row_ids(
     mapping
 }
 
+/// Remap a given index using the fragment reuse index if possible.
+/// If the frag reuse index does not exist, the operation fails with [Error::NotSupported]
+/// If the frag reuse index exists but is empty, the operation succeeds without a commit.
+async fn remap_index(dataset: &mut Dataset, index_id: &Uuid) -> Result<()> {
+    let indices = dataset.load_indices().await.unwrap();
+    let frag_reuse_index_meta = match indices.iter().find(|idx| idx.name == FRAG_REUSE_INDEX_NAME) {
+        None => Err(Error::NotSupported {
+            source: "Fragment reuse index not found, cannot remap an index post compaction".into(),
+            location: location!(),
+        }),
+        Some(frag_reuse_index_meta) => Ok(frag_reuse_index_meta),
+    }?;
+
+    let frag_reuse_details = load_frag_reuse_index_details(dataset, frag_reuse_index_meta)
+        .await
+        .unwrap();
+    let frag_reuse_index =
+        open_frag_reuse_index(frag_reuse_details.as_ref(), dataset.fragments().as_slice())
+            .await
+            .unwrap();
+
+    if frag_reuse_index.row_id_maps.is_empty() {
+        return Ok(());
+    }
+
+    // Sequentially apply the row id maps from oldest to latest
+    let mut curr_index_id = *index_id;
+    for (i, row_id_map) in frag_reuse_index.row_id_maps.iter().enumerate() {
```

Comment:
Is it possible to merge all these transactions into one?  Or is there a concern that might be too much work for one operation?  Or some other concern?

Reply:
I was mainly under the mindset of (1) preserving intermediate work in case anything goes wrong, (2) currently there is an assumption that the index to remap is already committed, and there are logic (e.g. remap_vector_index) that tries to fetch and open the index from the dataset. To introduce an uncommitted index concept it will require quite big refactoring effort. But I agree it would be nice to be able to just do this in a single transaction to minimize the number of transactions, I can create an issue to track that.